lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -0,0 +1,1009 @@
+//===- Dedup.cpp - FIRRTL module deduping -----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements FIRRTL module deduplication.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/SHA256.h"
+
+using namespace circt;
+using namespace firrtl;
+using hw::InnerRefAttr;
+
+/// This stores a mapping from a module name to every NLA that it particiapates
+/// in.
+using NLAMap = DenseMap<Attribute, std::vector<NonLocalAnchor>>;
```

Comment:
When we go through `addAnnotationContext` we will be cloning each NLA in the vector for every parent module.  When we create the new NLA, it needs to be recorded in the `NLAMap`, which can cause it to grow. Since the `NLAMap` grows while we're iterating the NLA vector stored in the map, we will crash if we use a `SmallVector` but we're safe with a `std::vector.` This is honestly a bit sketchy, there is some disagreement if moving a `std::vector` does not invalidate iterators, but this is done in a couple places over LLVM and a lot of things would break otherwise.